crates/ruff_python_semantic/src/model.rs

```diff
@@ -856,6 +865,27 @@ impl<'a> SemanticModel<'a> {
         &self.scopes[self.scope_id]
     }
 
+    /// Returns the parent of the given scope, if any.
+    pub fn scope_parent(&self, scope: &Scope) -> Option<&Scope<'a>> {
+        scope.parent.map(|scope_id| &self.scopes[scope_id])
+    }
+
+    /// Returns the first parent of the given scope that is not a [`ScopeKind::Type`] scope, if any.
+    pub fn scope_parent_skip_types(&self, scope: &Scope) -> Option<&Scope<'a>> {
+        let mut current_scope = scope;
+        loop {
+            if let Some(parent) = self.scope_parent(current_scope) {
+                if parent.kind.is_type() {
+                    current_scope = parent;
+                } else {
+                    return Some(parent);
+                }
+            } else {
+                return None;
+            }
+        }
```

Comment:
I think this can be written as a `while-let` loop:

```rust
/// Returns the first parent of the given scope that is not a [`ScopeKind::Type`] scope, if any.
pub fn scope_parent_skip_types(&self, scope: &Scope) -> Option<&Scope<'a>> {
    let mut current_scope = scope;
    while let Some(parent) = self.scope_parent(current_scope) {
        if parent.kind.is_type() {
            current_scope = parent;
        } else {
            return Some(parent);
        }
    }
    None
}
```

Reply:
Ah cool didn't know you could do that, will try.

Reply:
Or you use `std::iter::successors`

```rust
std::iter::successors(self.parent_scope(current_scope), |scope| self.parent_scope(scope)).find(|scope| !scope.kind.is_type()) 
```

Being able to iterate over all `ancestor_scopes` may be generally useful and would improve readability here:


```rust
pub fn ancestor_scopes(&self, scope: &Scope) -> impl Iterator<Item = &Scope> {
	std::iter::successors(self.scope_parent(scope), |next| self.scope_parent(next)) 
} 

pub fn first_value_scope(&self, scope: &Scope) -> Option<&Scope> {
	self.ancestor_scopes(scope).find(|scope| !scope.kind.is_type())
}
```

Reply:
FWIW we have `ancestors` on the `Scopes` struct, but it needs a `ScopeId` not a `Scope`, and we don't actually store the ID on `Scope`. So we'd need to refactor to pass the ID here rather than the `Scope` itself.

Reply:
Unsure how to proceed here. I'm not really sure how we want to deal with overlap on the `Scopes` struct.

Reply:
I think using a loop is fine for now, we can always consider this refactor later as it's somewhat orthogonal.